pkg/controller/pki/pki_controller.go

```diff
@@ -0,0 +1,265 @@
+		if apierrors.IsNotFound(err) {
+			log.Printf("PKI %s seems to have been deleted\n", request.NamespacedName)
+			return reconcile.Result{}, nil
+		}
+		log.Println(err)
+		return reconcile.Result{}, err
+	}
+
+	// Check to see if we already know this object
+	existing := r.pkis[request.NamespacedName]
+	if existing != nil {
+		// If the spec has changed, refresh
+		if !reflect.DeepEqual(obj.Spec, existing.spec) {
+			log.Printf("PKI %s has changed, refreshing\n", request.NamespacedName)
+			delete(r.pkis, request.NamespacedName)
+			existing = nil
+		}
+	}
+	if existing == nil {
+		existing, err = newPKI(obj, r.clientset, r.mgr)
+		if err != nil {
+			log.Println(err)
+			r.pkiErrs[request.NamespacedName] =
+				errors.Wrapf(err, "could not parse PKI.Spec %s", request.NamespacedName)
+			r.setStatus()
+			return reconcile.Result{}, err
+		}
+		r.pkis[request.NamespacedName] = existing
+	}
+
+	err = existing.sync()
+	if err != nil {
+		log.Println(err)
+		r.pkiErrs[request.NamespacedName] =
+			errors.Wrapf(err, "could not reconcile PKI %s", request.NamespacedName)
+		r.setStatus()
+		return reconcile.Result{}, err
+	}
+
+	log.Println("successful reconciliation")
+	delete(r.pkiErrs, request.NamespacedName)
+	r.setStatus()
+	return reconcile.Result{RequeueAfter: ResyncPeriod}, nil
+}
+
+// setStatus summarizes the status of all PKI objects and updates the statusmanager
+// as appropriate.
+func (r *PKIReconciler) setStatus() {
+	if len(r.pkiErrs) == 0 {
+		r.status.SetNotDegraded(statusmanager.PKIConfig)
+	} else {
+		msgs := []string{}
+		for _, e := range r.pkiErrs {
+			msgs = append(msgs, e.Error())
+		}
+		r.status.SetDegraded(statusmanager.PKIConfig, "PKIError", strings.Join(msgs, ", "))
+	}
+}
+
+// pki is the internal type that represents a single PKI CRD. It manages the
+// business of reconciling the certificate objects
+type pki struct {
+	spec       netopv1.PKISpec
+	controller *certrotation.CertRotationController
+	lastErr    error
+}
+
+// newPKI creates a CertRotationController for the supplied configuration
+func newPKI(config *netopv1.PKI, clientset *kubernetes.Clientset, mgr manager.Manager) (*pki, error) {
+	spec := config.Spec
+
+	// Ugly: the existing cache + informers used as part of the controller-manager
+	// can't be used, because they're untyped. So, we need to create our own.
+	// However, this has a few advantages - namely, we're creating a namespace-scoped
+	// watch, which is much more efficient than watching all Secrets and ConfigMaps
+	// TODO: consider adding a label selector to the watch, since we can do that.
+
+	inf := informers.NewSharedInformerFactoryWithOptions(
+		clientset,
+		24*time.Hour,
+		informers.WithNamespace(config.Namespace))
+
+	cont, err := certrotation.NewCertRotationController(
+		fmt.Sprintf("%s/%s", config.Namespace, config.Name),
+		certrotation.SigningRotation{
+			Namespace:     config.Namespace,
+			Name:          spec.CACert.Name,
+			Validity:      spec.CACert.ValidFor.Duration,
```

Comment:
I would not expose duration values to end users.  You'll get people at the crazy ends of the spectrum and this certrotationcontroller is likely to misbehave at the ends.

/cc @sttts

Reply:
Fair enough.